kotlin/flatbuffers-kotlin/src/commonMain/kotlin/com/google/flatbuffers/kotlin/JSON.kt

```diff
@@ -0,0 +1,788 @@
+    }
+    return skipWhitespace(limit) { data[it] }
+  }
+
+  private inline fun skipWhitespace(limit: Int, crossinline fetch: (Int) -> Byte): Byte {
+    var pos = readPos
+    while (pos < limit) {
+      val d = fetch(pos++)
+      if (!isWhitespace(d)) {
+        readPos = pos
+        return d
+      }
+    }
+    readPos = limit
+    return LETTER_EOF
+  }
+
+  /**
+   * Parse a json as [String] and returns a [Reference] to a FlexBuffer.
+   */
+  public fun parse(data: String): Reference = parse(ArrayReadBuffer(data.encodeToByteArray()))
+
+  /**
+   * Parse a json as [ByteArray] and returns a [Reference] to a FlexBuffer.
+   */
+  public fun parse(data: ByteArray): Reference = parse(ArrayReadBuffer(data))
+
+  /**
+   * Parse a json as [ReadBuffer] and returns a [Reference] to a FlexBuffer.
+   */
+  public fun parse(data: ReadBuffer): Reference {
+    reset()
+    parseValue(data, skipWhitespace(data), null)
+    if (readPos < data.limit) {
+      val tok = skipWhitespace(data)
+      if (tok != LETTER_EOF) {
+        makeError(data, "Extraneous charaters after parse has finished", tok)
+      }
+    }
+    output.finish()
+    return getRoot(output.buffer)
+  }
+
+  private fun parseValue(data: ReadBuffer, first: Byte, key: String? = null): FlexBufferType {
+    return when (first) {
+      LETTER_OPEN_OBJECT -> parseObject(data, key)
+      LETTER_OPEN_ARRAY -> parseArray(data, key)
+      LETTER_t -> parseTrue(data, key)
+      LETTER_n -> parseNull(data, key)
+      LETTER_f -> parseFalse(data, key)
+      LETTER_DOUBLE_QUOTE -> parseString(data, key)
+      LETTER_0, LETTER_1, LETTER_2, LETTER_3, LETTER_4, LETTER_5,
+      LETTER_6, LETTER_7, LETTER_8, LETTER_9, LETTER_MINUS -> parseNumber(data, data.data(), first, key)
+      else -> makeError(data, "Unexpected Character while parsing", first)
+    }
+  }
+
+  private fun parseObject(data: ReadBuffer, key: String? = null): FlexBufferType {
+    val fPos = output.startMap()
+    var lastTok = LETTER_OPEN_OBJECT
+    var tok: Byte = skipWhitespace(data)
+    val limit = data.limit
+    while (readPos <= limit) {
+      when (tok) {
+        LETTER_CLOSE_OBJECT -> {
+          if (lastTok == LETTER_COMMA) {
+            makeError(data, "Expecting a value after ${LETTER_COMMA.print()}", tok)
+          }
+          output.endMap(fPos, key); return T_MAP
+        }
+        LETTER_COMMA, 0.toByte() -> {
+          if (lastTok == LETTER_OPEN_OBJECT || lastTok == LETTER_COMMA) {
+            makeError(data, "Expecting a key/value after ${lastTok.print()}", tok)
+          }
+          lastTok = tok
+          tok = skipWhitespace(data)
+        }
+        LETTER_DOUBLE_QUOTE -> {
+          val childKey = readString(data)
+          tok = skipWhitespace(data)
+          if (tok.toChar() != ':')
+            makeError(data, "Expecting ':' after object Key", tok)
+          parseValue(data, skipWhitespace(data), childKey)
+          lastTok = tok
+          tok = skipWhitespace(data)
```

Comment:
This explicit skipping of whitespace in a parser is pretty fragile. Instead, have a separate lexical analyzer function that results in tokens, and can deal with whitespace centrally. See `Parser::Next()` in `idl_parser.cpp` for example. But up to you if you want to change that here or as a future improvement.

Reply:
I add a tokeniser, can you take a look?